rust/relay/ebpf-turn-router/src/main.rs

```diff
@@ -164,135 +164,661 @@ fn try_handle_turn_ipv4(ctx: &XdpContext, eth: Eth) -> Result<(), Error> {
+            .remove_u16(old_ipv4_len)
+            .add_u32(u32::from_be_bytes(new_ipv4_dst.octets()))
+            .add_u16(new_ipv4_len)
+            .into_checksum(),
     );
 
-    add_channel_data_header_ipv4(
-        ctx,
-        CdHdr {
-            number: channel_number.to_be_bytes(),
-            length: channel_data_length.to_be_bytes(),
-        },
-    )?;
+    //
+    // 3. UDP header
+    //
+    let new_udp_src = port_and_peer.allocation_port();
+    let new_udp_dst = port_and_peer.peer_port();
+    let new_udp_len = old_udp_len - CdHdr::LEN as u16;
+
+    let udp_ptr =
+    // SAFETY: After packet shrinking, the offset is valid within adjusted bounds.
+        unsafe { ref_mut_at::<UdpHdr>(ctx, NET_SHRINK as usize + EthHdr::LEN + Ipv4Hdr::LEN)? };
+    (*udp_ptr).set_source(new_udp_src);
+    (*udp_ptr).set_dest(new_udp_dst);
+    (*udp_ptr).set_len(new_udp_len);
+
+    // Incrementally update UDP checksum
+
+    // TODO: Remove conditional checksums once we can test this fully in CI
+    if old_udp_check == 0 || !crate::config::udp_checksum_enabled() {
+        // No checksum is valid for UDP IPv4 - we didn't write it, but maybe a middlebox did
+        (*udp_ptr).set_check(0);
+    } else {
+        let check = ChecksumUpdate::new(old_udp_check)
+            .remove_u32(u32::from_be_bytes(old_ipv4_src.octets()))
+            .remove_u16(old_udp_src)
+            .remove_u16(old_udp_dst)
+            .remove_u16(old_udp_len)
+            .remove_u16(old_udp_len)
+            .remove_u16(channel_number)
+            .remove_u16(channel_data_length)
+            .add_u32(u32::from_be_bytes(new_ipv4_dst.octets()))
+            .add_u16(new_udp_src)
+            .add_u16(new_udp_dst)
+            .add_u16(new_udp_len)
+            .add_u16(new_udp_len)
+            .into_checksum();
+
+        if check == 0 {
+            (*udp_ptr).set_check(0xFFFF); // Special case for zero checksum - write 0xFFFF to the wire
+        } else {
+            (*udp_ptr).set_check(check);
+        }
+    }
+
+    // Now shrink the packet by removing the first 4 bytes
+    // SAFETY: bpf_xdp_adjust_head is a kernel helper that safely adjusts packet headroom.
+    let ret = unsafe { bpf_xdp_adjust_head(ctx.ctx, NET_SHRINK) };
+    if ret < 0 {
+        return Err(Error::XdpAdjustHeadFailed(ret));
+    }
 
     Ok(())
 }
 
+// Convert IPv4 to IPv6 and remove channel data header
 #[inline(always)]
-fn try_handle_turn_ipv6(ctx: &XdpContext, eth: Eth) -> Result<(), Error> {
+fn handle_ipv4_to_ipv6_channel_to_udp(
+    ctx: &XdpContext,
+    port_and_peer: &PortAndPeerV6,
+) -> Result<(), Error> {
+    // Expand FIRST - 20 - 40 + 4 = -16 bytes
+    const NET_EXPANSION: i32 = Ipv4Hdr::LEN as i32 - Ipv6Hdr::LEN as i32 + CdHdr::LEN as i32;
+
+    // SAFETY: bpf_xdp_adjust_head is a kernel helper that safely adjusts packet headroom.
+    let ret = unsafe { bpf_xdp_adjust_head(ctx.ctx, NET_EXPANSION) };
```

Comment:
Might be worth introducing our own wrapper here so you can just call it and `?` the error.

Reply:
Good call, I'll update this